File: lasvecka-python/calc_date.py
from datetime import date, timedelta
import json

EASTER_START = date.fromisoformat("2021-04-01")
EASTER_END = date.fromisoformat("2021-04-10")
ORD_CONT = date.fromisoformat("2021-04-12")

# Find the appropiate date & period from data.txt (the output from the webscraper)
def read_date_period(curr_date):
    sought_date = ""
    diff = -1000
    with open('data.txt', 'r') as data_file:
        date_dict = eval(data_file.read())
        for dat, typ in date_dict.items():
            delta_t = date.fromisoformat(dat) - curr_date
            if delta_t.days == 0:
                return (dat, typ)
            elif delta_t.days > diff and delta_t.days < 0:
                sought_date = dat
                diff = delta_t.days
        return(sought_date, date_dict[sought_date])




def handle_easter(easter_start_diff, easter_end_diff):
    # Checks if it's easter
    if easter_start_diff.days >= 0 and easter_end_diff.days <= 0:
        return("Självstudier")
    # Add offset, will go up to Lv 8 until regular exam period
    elif easter_end_diff.days > 0:
        weeks, _ = divmod(easter_end_diff.days, 7)
        return ("Lv " + str(weeks + 2))
    


# Computes and determines the week of the study period, or returns Självstudier/Tentavecka if it's Easter or
# exam period.
def compute_time():
    current_date = date.fromisoformat("2021-04-28")
    # Calculates diffs for easter handling
    easter_end_check =  current_date - ORD_CONT
    easter_start_check = current_date - EASTER_START
    dat, typ = read_date_period(current_date)
    print(dat, typ)
    # Is it exam period? Also check for final exam period
    if typ == "exam_period":
        print(type(dat))
        delta_t = current_date - date.fromisoformat(dat)
        if delta_t.days > 7:
            return "Självstudier"
        else:
            return "Tentavecka"
    # Is it easter or after easter? Call handle_easter
    if easter_end_check.days >= 0 or easter_start_check.days >= 0:
        return (handle_easter(easter_start_check, easter_end_check))
    else:
        # Regular time handling
        delta_t = current_date - date.fromisoformat(dat)
        weeks, _ = divmod(delta_t.days, 7)
        if weeks > 8:
            return ("Självstudier")
        else:
            return ("LV " + str(weeks))

# For testing purposes
if __name__ == '__main__':
    print(compute_time())

File: lasvecka-node/calc_date.js
const moment = require("moment");
const fs = require("node:fs");
const scrape = require("./lasveckor_scraper.js");
let dateDict = {};

// check if data.json exists
if (!fs.existsSync("./data/data.json")) {
	scrape().then((res) => {
		dateDict = res;
	});
} else {
	dateDict = JSON.parse(fs.readFileSync("./data/data.json"));

	// check if it was updated after 1/7 this year and if 1/7 has occured this year
	const updated = moment(dateDict.updated);
	const firstOfJuly = moment().month(6).date(1);
	if (updated.isBefore(firstOfJuly) && moment().isAfter(firstOfJuly)) {
		scrape().then((res) => {
			dateDict = res;
		});
	}
}

function readDatePeriod(currDate) {
	let soughtDate = "";
	let diff = -1000;
	for (const dat in dateDict) {
		// if dat is updated, easter_start, easter_end or ord_cont then skip
		if (
			dat === "updated" ||
			dat === "easter_start" ||
			dat === "easter_end" ||
			dat === "ord_cont"
		) {
			continue;
		}

		const deltaT = moment(dat).diff(currDate, "days");
		if (deltaT === 0) {
			return { date: dat, type: dateDict[dat] };
		}

		if (deltaT > diff && deltaT < 0) {
			soughtDate = dat;
			diff = deltaT;
		}
	}
	return { date: soughtDate, type: dateDict[soughtDate] };
}

function computeTime() {
	// Find date where value is easter_start, easter_end and ord_cont in json file
	const EASTER_START = dateDict.easter_start;
	const ORD_CONT = dateDict.ord_cont;
	const currentDate = moment();
	const easterEndCheck = currentDate.diff(ORD_CONT, "days");
	const easterStartCheck = currentDate.diff(EASTER_START, "days");
	const { date: dat, type: typ } = readDatePeriod(currentDate);
	if (typ === "exam_period") {
		const deltaT = currentDate.diff(moment(dat), "days");
		if (deltaT > 7) {
			return "Självstudier";
		}
		return "Tentavecka";
	}

	let deltaT = currentDate.diff(moment(dat), "days");
	if (typ === 'first_day') {
		const weeks = Math.floor(deltaT / 7);
		return `MV ${weeks + 1}`;
	}

	if (easterStartCheck >= 0 && easterEndCheck <= 0) {
		return "Självstudier";
	}

	if (easterEndCheck > 0) {
		// the easter period is over and it is 7 days long
		deltaT = deltaT - 7;
	}

	const weeks = Math.floor(deltaT / 7);

	if (weeks > 7) {
		return "Självstudier";
	}
	return `LV ${weeks + 1}`;
}

module.exports = computeTime;
